### PR TITLE
added signing of AccountData to ValidatorSigner

### PR DESCRIPTION
**Pull request by pompon0**
and adjusted the accounts_data::Cache accordingly.

### CODE REVIEW

File: chain/network/src/network_protocol/mod.rs [hunk 1]
```diff
@@ -52,7 +53,14 @@ pub struct AccountData {
...
 impl AccountData {
-    pub fn sign(self, signer: &dyn near_crypto::Signer) -> anyhow::Result<SignedAccountData> {
+    pub fn sign(self, signer: &dyn ValidatorSigner) -> anyhow::Result<SignedAccountData> {
+        if &self.account_id != signer.validator_id() {
+            anyhow::bail!(
+                "AccountData.account_id doesn't match the signer's account_id: got {}, want {}",
+                self.account_id,
+                signer.validator_id()
+            );
+        }
```
Comment (matklad) [inline chain/network/src/network_protocol/mod.rs:63]: Should this be an assert? This feels like a clear deterministic programming error.

File: core/primitives/src/validator_signer.rs [hunk 2]
```diff
@@ -49,6 +49,19 @@ pub trait ValidatorSigner: Sync + Send {
...
+    /// Making it typesafe would require moving the definition of
+    /// AccountKeyPayload proto to this crate to avoid a dependency cycle,
+    /// so for now we are just signing an already-serialized byte sequence.
+    /// We are serializing a proto rather than borsh here (as an experiment,
+    /// to allow the network protocol to evolve faster than on-chain stuff),
+    /// but we can always revert that decision, because these signatures are
+    /// used only for networking purposes and are not persisted on chain.
+    /// Moving to proto serialization for stuff stored on chain would be way
+    /// harder.
+    fn sign_account_key_payload(&self, proto_bytes: &[u8]) -> Signature;
```
Comment (matklad) [inline core/primitives/src/validator_signer.rs:63]: Thanks for providing a lot of context here, very helpful!

File: chain/network/src/network_protocol/mod.rs (hunk 1 above)
Comment (pompon0) [reply chain/network/src/network_protocol/mod.rs:63]: It is likely to be a programming error, but it should be the caller who should do the panic (via unwrap).

Comment (pompon0) [reply chain/network/src/network_protocol/mod.rs:63]: nvm, that would make sense if there were no valid runtime errors. I'll make it an assert.

Comment (matklad) [reply chain/network/src/network_protocol/mod.rs:63]: >It is likely to be a programming error, but it should be the caller who should do the panic (via unwrap).

Yeah, if this was a single error condition, and if we didn't use `anyhow`, that indeed would be better.

Comment (pompon0) [reply chain/network/src/network_protocol/mod.rs]: done

File: chain/network/src/network_protocol/mod.rs [hunk 3]
```diff
@@ -52,7 +53,21 @@ pub struct AccountData {
 pub const MAX_ACCOUNT_DATA_SIZE_BYTES: usize = 10000; // 10kB
 
 impl AccountData {
-    pub fn sign(self, signer: &dyn near_crypto::Signer) -> anyhow::Result<SignedAccountData> {
+    /// Serializes AccountData to proto and signs it using `signer`.
+    /// Panics if AccountData.account_id doesn't match signer.validator_id(),
+    /// as this would likely be a bug.
+    /// Returns an error if the serialized data is too large to be broadcasted.
+    /// TODO(gprusak): consider separating serialization from signing (so introducing an
+    /// intermediate SerializedAccountData type) so that sign() then could fail only
```
Comment (matklad) [inline chain/network/src/network_protocol/mod.rs:61]: I always view failable serialization with suspicion: if serialization is failable, it raises the question of why we are alowing such potentially invalid values to exist in the first place? 

In this case, I'd probably try to enforce size limit as an invariant of `AccountData` itself (so, eg, make it hash <= 10 peers or something)

But yeah, `serialize()?.sign()?` is also a great general approach!

**Review by matklad — APPROVED** [review]

File: chain/network/src/network_protocol/mod.rs (hunk 3 above)
Comment (pompon0) [reply chain/network/src/network_protocol/mod.rs:61]: The AccountData is POD struct (i.e. with public fields) and it acts more like a builder. I've put a constraint on the total size of the serialized thing, which is a constraint verifiable once the value is "built" and I don't see anything wrong with that. Here sign() call acts as a constructor of SignedAccountData struct which is more than just a composition of its arguments (i.e. has additional constraints).

**Pull request merged**